contrib/vcl/source/vcl_io_handle.cc

```diff
@@ -0,0 +1,768 @@
+                                         unsigned long*) {
+  return {0, 0};
+}
+
+Api::SysCallIntResult VclIoHandle::setBlocking(bool) {
+  uint32_t flags = O_NONBLOCK;
+  uint32_t buflen = sizeof(flags);
+  int32_t rv = vppcom_session_attr(sh_, VPPCOM_ATTR_SET_FLAGS, &flags, &buflen);
+  return {rv < 0 ? -1 : 0, -rv};
+}
+
+absl::optional<int> VclIoHandle::domain() {
+  VCL_LOG("grabbing domain sh {:x}", sh_);
+  return {AF_INET};
+};
+
+Envoy::Network::Address::InstanceConstSharedPtr VclIoHandle::localAddress() {
+  vppcom_endpt_t ep;
+  uint32_t eplen = sizeof(ep);
+  uint8_t addr_buf[sizeof(struct sockaddr_in6)];
+  ep.ip = addr_buf;
+  if (vppcom_session_attr(sh_, VPPCOM_ATTR_GET_LCL_ADDR, &ep, &eplen)) {
+    return nullptr;
+  }
+  return vclEndptToAddress(ep, sh_);
+}
+
+Envoy::Network::Address::InstanceConstSharedPtr VclIoHandle::peerAddress() {
+  VCL_LOG("grabbing peer address sh {:x}", sh_);
+  vppcom_endpt_t ep;
+  uint32_t eplen = sizeof(ep);
+  uint8_t addr_buf[sizeof(struct sockaddr_in6)];
+  ep.ip = addr_buf;
+  if (vppcom_session_attr(sh_, VPPCOM_ATTR_GET_PEER_ADDR, &ep, &eplen)) {
+    return nullptr;
+  }
+  return vclEndptToAddress(ep, sh_);
+}
+
+void VclIoHandle::updateEvents(uint32_t events) {
+  auto wrk_index = vclWrkIndexOrRegister();
+  VclIoHandle* vcl_handle = this;
+
+  if (wrk_index && is_listener_) {
+    vcl_handle = wrk_listener_.get();
+  }
+
+  struct epoll_event ev;
+  ev.events = EPOLLET;
+
+  if (events & Event::FileReadyType::Read) {
+    ev.events |= EPOLLIN;
+  }
+  if (events & Event::FileReadyType::Write) {
+    ev.events |= EPOLLOUT;
+  }
+  if (events & Event::FileReadyType::Closed) {
+    ev.events |= EPOLLERR | EPOLLHUP;
+  }
+
+  ev.data.u64 = reinterpret_cast<uint64_t>(vcl_handle);
+
+  vppcom_epoll_ctl(vclEpollHandle(wrk_index), EPOLL_CTL_MOD, vcl_handle->sh(), &ev);
+}
+
+void VclIoHandle::initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                      Event::FileTriggerType, uint32_t events) {
+  VCL_LOG("adding events for sh {:x} fd {} isListener {}", sh_, fd_, isListener());
+
+  auto wrk_index = vclWrkIndexOrRegister();
+  vclInterfaceRegisterEpollEvent(dispatcher);
+
+  VclIoHandle* vcl_handle = this;
+
+  if (is_listener_) {
+    if (wrk_index) {
+      // If this is not the main worker, make sure a worker listener exists
+      if (!wrk_listener_) {
+        vppcom_endpt_t ep;
+        uint8_t addr_buf[sizeof(struct sockaddr_in6)];
+        ep.ip = addr_buf;
+        uint32_t proto;
+
+        if (peekVclSession(sh_, &ep, &proto)) {
+          RELEASE_ASSERT(0, "");
+        }
+
+        auto address = vclEndptToAddress(ep, -1);
+        auto sh = vppcom_session_create(proto, 1);
+        wrk_listener_ = std::make_unique<VclIoHandle>(static_cast<uint32_t>(sh), VclInvalidFd);
+        wrk_listener_->bind(address);
+        uint32_t rv = vppcom_session_listen(sh, 5);
```

Comment:
Better replace 5 with a constant with a doc string.

Reply:
This is actually ignored (the api is posix-like but vcl has opinions about the second value). So if it not an issue, I'll just add a comment here.